AkiraMD/ruby-senior-take-home-assignment
# PG/interview/ruby take home

## Overview
This PR contains the following changes as per the requirements:

* `/patients/:patient_id` fetches single patient with all possible error handlings.
<img width="572" alt="Screenshot 2024-05-30 at 8 49 49 AM" src="https://github.com/AkiraMD/ruby-senior-take-home-assignment/assets/2009988/35ec4261-6270-461d-a9ca-dfa6b4a5d59c">
* `/patients/:patient_id/record` fetches single patient's records from external vendor APIs.
<img width="568" alt="Screenshot 2024-05-30 at 8 50 33 AM" src="https://github.com/AkiraMD/ruby-senior-take-home-assignment/assets/2009988/f221fa59-e535-41c3-9152-37f0b8162323">

* Cache implementation for caching the response of `/patients/:patient_id/record`
* Rspec test suite for the REST services and Caching.
<img width="898" alt="Screenshot 2024-05-30 at 8 51 27 AM" src="https://github.com/AkiraMD/ruby-senior-take-home-assignment/assets/2009988/b5162de2-601d-432d-830d-a88796b75ba2">


## How to test
`docker compose up --build` to build the project from scratch and then we can use the browser or an API client ( like insomnia or postman) to test the APIs.

To run the specs, perform the following:
*  Ensure docker is running.
*  Run `docker exec -it rest_server bash` in a terminal. This will log you in the rest server terminal.
*  Run `bundle exec rspec spec/`

## Challenges
* We wrote our own business layer logic (the models) where we are communicating with raw SQL. If performance is not an issue, we can use something like an ActiveRecord to handle all business layer logic. This will helps us have better separation with the DB and also provides measures for N + 1 like scenarios.
* There was no hot-reload for API changes. Whenever I was making a change in the API, I was restarting the docker to see my changes. We can think of have something like that to help developer productivity.
* There was a bug in `docker-compose.yml` for mock api 2. Had to resolve it for successful implementation.

## What I could have done better
* Could have implemented the `Vandelay/Integrations/Vendor/Base` in a better way. Current implementation is very rudimentary and doesn't handle failures.
* I wrote a separate REST implementation for `patients/:patient_id/record` as I failed to implement it as a nested resource in `patient/:patient_id`. If I could have implemented it in the single patient resource fetch somehow, then that would have resulted in more DRY code.
